src/addrman.cpp

```diff
@@ -713,37 +713,49 @@ CAddrInfo CAddrMan::Select_(bool newOnly) const
         // use a tried node
         double fChanceFactor = 1.0;
         while (1) {
+            // Pick a tried bucket, and an initial position in that bucket.
             int nKBucket = insecure_rand.randrange(ADDRMAN_TRIED_BUCKET_COUNT);
             int nKBucketPos = insecure_rand.randrange(ADDRMAN_BUCKET_SIZE);
-            while (vvTried[nKBucket][nKBucketPos] == -1) {
-                nKBucket = (nKBucket + insecure_rand.randbits(ADDRMAN_TRIED_BUCKET_COUNT_LOG2)) % ADDRMAN_TRIED_BUCKET_COUNT;
-                nKBucketPos = (nKBucketPos + insecure_rand.randbits(ADDRMAN_BUCKET_SIZE_LOG2)) % ADDRMAN_BUCKET_SIZE;
+            // Iterate over the positions of that bucket.
+            for (int i = 0; i < ADDRMAN_BUCKET_SIZE; ++i) {
+                if (vvTried[nKBucket][nKBucketPos] != -1) {
+                    int nId = vvTried[nKBucket][nKBucketPos];
+                    const auto it_found{mapInfo.find(nId)};
+                    assert(it_found != mapInfo.end());
+                    const CAddrInfo& info{it_found->second};
+                    if (insecure_rand.randbits(30) < fChanceFactor * info.GetChance() * (1 << 30)) return info;
```

Comment:
nit: while you touching this, we could also make this more readable by avoiding bit operations :)

Reply:
Any suggestions?

Reply:
I meant `insecure_rand.randrange(XYZ) < fChanceFactor * info.GetChance() * XYZ` should just work?